### ISSUE TITLE
Allow specifying import order of global styles

### ISSUE THREAD

**Issue by max** [salsify/ember-css-modules#7]
It is common to have a set of base styles for a given application. Those base styles often rely on import order. By default `ember-cli` takes care of this:

> In the production build, the `@import` statements are replaced with the contents of their files and the final minified, concatenated single CSS file is built to `dist/assets/yourappname-FINGERPRINT_GOES_HERE.css`.

An example in Ember land:

```css
/* app/styles/app.css */

@import "normalize.css";
@import "base.css";
@import "typography.css";
...
@import "utilities.css";
```

When installing `ember-css-modules` this no longer works. Instead all files in `app/styles/` are concatenated alphabetically – which removes the ability to specify import order.

Besides basic functionality it would also be fantastic if `postcss-import` was supported. This allows globbing, importing NPM modules, etc. (in the above example you could imagine `normalize.css` just being an NPM module vs vendored in).

**Comment by dfreeman** [comment]
@max based on our conversation earlier, I've been rolling around the different parts of this in my head and thinking about possible approaches. A couple questions:
 - for stuff like `normalize.css`, I've been using [ember-cli-node-assets](https://github.com/dfreeman/ember-cli-node-assets) to just include styles from node modules in the vendor stylesheet – would that solve your "pull stuff from NPM" needs, or are there cases where you need finer-grained control relative to app imports?
 - this may end up being immaterial, but do you see a need to be able to import things outside of the tree of files `@import`ed from app.css? In other words, would it work for the final output to be something like this?

   ```
   <topo-sorted contents of things @imported directly or indirectly from app.css>
   <contents of app.css>
   <arbitrarily-ordered contents of all the rest of the CSS files in your app tree>
   ```

**Comment by max** [comment]
- I will give ember-cli-node-assets a try. I'm not sure how it works but I often want to import an NPM module like [basscss](http://www.basscss.com/) in `app.css` and then customize some variables (custom properties) to tweak defaults. I'm guessing I can't do that if it's in `vendor.css` but I'll double check.
- If by the "the rest" you mean all styles that are not explicitly imported in `app.css` that's fine. I generally don't have styles that are not imported through app.css; it's a non issue. I wonder if you could just detect if there is an `app.css` present and then only import that. If there is no `app.css` import all in random order.

**Comment by dfreeman** [comment]
@max I've got [a branch](https://github.com/salsify/ember-css-modules/tree/configure-concatenation) that allows for finer-grained configuration of the concatenation process (most of that work was going to be needed for eventual better sourcemap support anyway), as well as exposing the ability to load modules to plugins.

I also threw together a [proof-of-concept app](https://github.com/dfreeman/css-modules-and-imports) that makes use of that branch. It [does the following](https://github.com/dfreeman/css-modules-and-imports/commit/b9abe827a7da470f9023d88870c8969bacd9c491#diff-7):
 - pulls in basscss via ember-cli-node-assets to the `vendor.css` file
 - boosts `application.css` to be the first file concatenated into the output app CSS
 - tells the addon not to automatically concatenated anything out of the `styles/base` directory
 - explicitly `@import`s two files from that directory via `postcss-imports`

Since the vendor stylesheet is always included before the app one, overriding `basscss` is easy. For as simple a case as this example, you could also just forego postcss-imports entirely just add the files in base directly to the `headerFiles` array, but for more complex situation the flexibility of the plugin is probably nice.

**Comment by dfreeman** [comment]
Re your "the rest" comment – explicitly `@import`ing _all_ CSS files in your app feels like an antipattern to me, since the point of CSS modules is to help eliminate coupling between different parts of your stylesheet.
For base-level stuff it makes sense to pull things to the top, but if I ever found myself needing to explicitly include the styles for a particular component or controller before or after another, that seems like a smell.

**Comment by max** [comment]
Wow, fantastic. Going to give this a try tonight and report back.

- Basscss wouldn't be overrides as much as changing custom properties before it gets processed. I don't think it's an issue with your solution.
- Misunderstood what you meant with "the rest". I agree what you described is an anti-pattern. Personally I don't think I would put anything but base styles in `app/styles`. In my mind `app/styles` should either contain global stuff (resets, typography, base elements) and utilities that never get overridden.

Thanks for looking into this. I'll let you know how it goes with the branch.

**Comment by dfreeman** [comment]
Gotcha. For pod structured apps (which is my absolute preference), I'd definitely never put anything other than base stuff in `app/styles`. For "classic" structured apps, though, `styles` ends up being a root entity directory like `controllers` or `templates`, so you can end up with controller/component-specific styles in there in parallel with your app structure.

Let me know how the branch works out! Now that I understand your use case slightly better, I'm guessing you probably won't want [the custom load function](https://github.com/dfreeman/css-modules-and-imports/commit/b9abe827a7da470f9023d88870c8969bacd9c491#diff-2fe2a7faee7de1e0d5d82e032498d84cR13) for `postcss-import`, since you're only `@import`ing global stuff.

**Comment by max** [comment]
Finally got around to trying this out and it's already so much closer to something we could use in production at @usecanvas. Thanks for your work. Here are some additional thoughts:

## Convention over configuration

Currently you require configuration to set `headerFiles` and you require explicit exclusion for files that should not be automatically included (i.e. `exclude: [...]`. I understand this decision but it seems somewhat at odds with Ember's convention over configuration.

I think it may be possible to make away with this configuration by sticking to some established conventions.

A new Ember app by default does not automatically concatenate styles in `app/styles`. They have to be explicitly imported in the seeded `app.css` file. Users are therefor already familiar with having to import styles. This add-on could behave similarly.

If we also use pods as a forcing function to organize stylesheets – similarly to how glimmer components will introduce breaking changes in a clean way – we can by default have `ember-css-modules` ignore the styles directory and only prefix pod styles. That would make new projects look like this:

```
app
├── components
│   └── my-button
│       ├── component.js
│       ├── styles.css
│       └── template.hbs
└── styles
    ├── app.css
    ├── base
    │   └── typography.css
    └── some-globals.css
```

Should a user then want to also prefix a folder within `app/styles` this could be done through configuration (`include: [...]`).

Switching from an exclude to an include for app styles and default to a pod structure for styles might seem aggressive but I think is more forward looking.

## Imports

I tried using `ember-cli-node-assets` to import Basscss but this does not allow for configuration of custom properties that are in the Basscss. For example:

```css
/* basscss/font-scale.css */
:root {
  --h1: 2rem;
}

.h1 {
  font-size: var(--h1);
}

/* ember-app/app/styles/app.css */
@import "normalize.css";
@import "basscss";

...

:root {
  --h1: 4rem;
}
```

To experiment I tried importing both Basscss and normalize.css with the `postcss-import` that now works and besides the fact that all Basscss classes were `ember-css-modules`-arized it seemed to work fine.

If we either switch from excludes to includes (as mentioned above) or I can exclude something like Basscss from being prefixed **Imports** would work as expected.

**Comment by max** [comment]
Oh, you could also just modularize either pod styles or styles in `app/styles/components`. This way non-pod structure apps still get the benefit of this addon without configuration. If anyone deviates from this convention they can then include additional folders to be prefixed.

**Comment by dfreeman** [comment]
Unfortunately route styles (i.e. those associated with a controller) can fall anywhere in the `styles` directory in a classic-structure app, so restricting to `styles/components` isn't feasible. On the `includes` vs `excludes` front, I don't think having to manually specify either one is a solid path forward in terms of being able to tell a good story about building a ground-up application using this addon. The current state of the branch is more a proof of concept than a candidate for being the final interface.

On the concatenation topic, I actually feel pretty strongly that concatenating **all** CSS that's anywhere in the `app` tree by default is important, but it's taken the better part of the past day to draw out into words the threads of reasoning behind that conviction. Ultimately it comes down to the confluence of several different factors:
 1. Any JavaScript file or HTMLbars template in the app tree is always automatically included in final application assets, regardless of whether they're directly imported anywhere; CSS has is actually kind of the oddball on that front.

 2. In "the olden days" the order you included JS assets on a page mattered – you needed to have your `jquery.js` script tag before your `jquery.supercoolplugin.js` script tag, or things would go sour. With the advent of JS modules, that ordering became much less important, except when dealing with global state, which is the exception rather than the rule. I see a direct parallel from there to CSS modules.

 3. On reflection, we've been talking about using `@import` in two very different capacities without clear differentiation. The standard CSS `@import` directive is more akin to the traditional C preprocessor `#include` than it is to ES6 `import` in that it just blindly injects the contents of one file into another.
  - For something like `basscss`, that blind injection is exactly what you want, in combination with additional processing steps afterward in order to tie together e.g. custom property usage
  - For a CSS modules, though, it's not actually clear to me what it should even mean to `@import` one module into another – the whole point is to give finer-grained control rather than just airlifting entire chunks of code around.

Stemming from each of those points, I see corresponding requirements that will need to be taken into account in any potential solution to the issues raised in this thread:
 1. Stylesheets should be elevated to the same status as JS modules and templates in that **if a CSS file in the app tree, it's in the final build**. Requiring manual inclusion of files represents a configuration burden on the user that should be unnecessary in the face of simple convention.

 2. Just like the vast majority of JS modules don't care what order they're concatenated into the final output, the vast majority of CSS modules also shouldn't care. However, appropriate tools need to be made available to the user to enforce ordering guarantees _in source_ when necessary. Just like I can `import 'foo/bar';` in JS to force execution of a module body only for side effects, **it must be possible within a CSS module to ensure that another module is included before it**. Crucially, I think `@import` is _not_ the right tool for this, since that already has a different well-defined meaning.

 3. There are benefits both during development (rebuild times) and in production (asset caching) to keeping vendor styling in `vendor.css` whenever possible. However, there are cases (such as framework configuration) where including vendor code into the app stylesheet is necessary, so **it must be possible to import 3rd party stylesheets into a module and process them against the same set of PostCSS plugins as app modules** (other than the modularization plugins themselves). For this use case, `@import` is likely appropriate.

I think a solution that meets the above three criteria would both handle your use case and be reasonable from a design and learnability standpoint.

For the second bullet, I imagine some other at-rule like `@require` or `@supersede` would be reasonable, though I'm not in love with either of those two names.
For the third, I think I'm going to have to bake `postcss-import` into the addon. My original goal with this addon had been not to force the use of any plugins other than the modularization suite, but my actual use  of it so far had only been with component-oriented addons, so I hadn't run up against these pain points with defining base application styles.

Apologies for the wall of text – this ended up being a little more of a novel than I'd anticipated. I have a rough implementation in mind for most of this, but haven't fully vetted the feasibility yet, and probably won't actually get to it until the weekend. If you've got further thoughts between now and then, I'm happy to hear them.

**Comment by max** [comment]
Thanks for the reply. Sorry for not getting back to you earlier. I can see you feel strongly about the automagic concatenation. If there is a way to work around that by either excluding or `@import`-ing it would fit my needs. I'm weary of introducing another concept like `@require` but it would get the job done.

**Comment by dfreeman** [comment]
My worry with using `@import` to tackle both module ordering and bringing in 3rd-party files is that, aside from being two different tasks, there are also two pretty different sets of technical constraints around how you'd want (and how it's possible) for each to function.
It seems likely to be confusing to users why some things might work when importing an outside file but not a module, or vice versa. I can go into a lot more detail on that, but I've already talked your ear off, so unless you're particularly interested I'll leave it there.

Anyway, I've updated the work in progress branch, as well as the [proof of concept app](https://github.com/dfreeman/css-modules-and-imports/commit/2962a317fca5e11b7af8fbd55f7ca5985a044b2c). I need to think some more about how to best present this functionality in the README and how to test it, but I think the actual implementation should be stable enough to play around with. You can:
 - [explicitly order modules](https://github.com/dfreeman/css-modules-and-imports/commit/2962a317fca5e11b7af8fbd55f7ca5985a044b2c#diff-3) in the final output using `@after-module` – I still don't love the name; I also played with `@prepend` and a couple others, but I don't feel like anything has quite captured the intent yet
 - [`@import` outside files into a module](https://github.com/dfreeman/css-modules-and-imports/commit/2962a317fca5e11b7af8fbd55f7ca5985a044b2c#diff-5), and they'll participate along with the rest of that module in any other PostCSS plugins that are configured (like `postcss-css-variables`), though without namespacing the imported classes

I really appreciate you iterating with me on this – it's super valuable to have an outside set of eyes on everything, and I think we're getting close to something that can provide a good balance of "just-works" convention in most cases while still being flexible enough to meet a lot of more specific needs.

**Comment by dfreeman** [comment]
@max I just published 0.3.0, which I believe covers the use cases here. If you have ordering goals you're still unable to accomplish with the addon, feel free to reopen or file a separate issue. Thanks again for bearing with me through this one.

**Comment by max** [comment]
@dfreeman Thanks! I've been OOO (Ember conf). Can't wait to dive in and use it in one of our products.

**Comment by BryanCrotaz** [comment]
I've just found this after days of struggling to work out why my sass was failing. 

Why not just use `@import` instead of `@after-module`? Its meaning is very clear and everyone knows how to use it already. And PLEASE add @after-module to the docs - the natural assumption is that sass will just work, and it doesn't without this hidden feature.
